### ISSUE TITLE
HomSap/AmericanAdmixture_4B11 is incompatible with discrete generations

### ISSUE THREAD

**Issue by grahamgower** [popsim-consortium/stdpopsim#378]
The Browning et al. 2011 American Admixture demographic model has faithfully reproduced a quirk of the originally published model---a small time delta has been added to events to ensure ordering. This works fine if time is in real valued coalescent units, but is not appropriate with integral valued generations in SLiM. In particular, the admixture proportions requested below will multiply to produce the *actual* desired proportions (mentioned in the comment) when ordering is guaranteed. However, conversion of the times to integers means these events all occur in the same  generation, and the intended proportions are not obtained.
```
    # Admixture event, 1/6 Africa, 2/6 Europe, 3/6 Asia
    admixture_event = [
        msprime.MassMigration(
            time=Tadmix, source=3, destination=0, proportion=1.0/6.0),
        msprime.MassMigration(
            time=Tadmix+0.0001, source=3, destination=1, proportion=2.0/5.0),
        msprime.MassMigration(
            time=Tadmix+0.0002, source=3, destination=2, proportion=1.0)
    ]
```

Probably the weird time deltas need to go away. Also see #62.

**Comment by jeromekelleher** [comment]
Hmmm, sounds quirky all right. FWIW, msprime does guarantee that events happen in the order they are supplied, even if they are at exactly the same time value. I don't see any reason for the time deltas, since we're assuming nothing happens in between anyway (right?)

**Comment by ndukler** [comment]
When I did the QC implementation I originally implemented it without the deltas and, to my memory, the order was correct. However, the original implementation by the paper's authors, and thus the port of that code over to stdpopsim used the deltas. We agreed to go with that at the time since it exactly matched the original implementation even if there were no real consequences. There should be no issue using integer valued generations.

**Comment by grahamgower** [comment]
Maybe I'm confused. Are the proportions given in the MassMigration events required to be different to the desired admixture proportions listed in the comments? I don't really understand how the desired proportions of {1/6, 2/6, 3/6} are reached.

**Comment by grahamgower** [comment]
Is this correct: `proportion=2.0/5.0`? Should it actually be `2.0/6.0`?

**Comment by ndukler** [comment]
Yes, that should be different because they are applied sequentially. Them together in the order they are applied to get the numbers for the admixture proportions.

**Comment by grahamgower** [comment]
I'll try to be as detailed as I can here, to explain why I think this model is problematic with SLiM. To recap the model, we have:
```
populations = [
        stdpopsim.Population(id="AFR", description="Contemporary African population"),
        stdpopsim.Population(id="EUR", description="Contemporary European population"),
        stdpopsim.Population(id="ASIA", description="Contemporary Asian population"),
        stdpopsim.Population(
            id="ADMIX", description="Modern admixed population"),
    ]
```
and
```
    # Admixture event, 1/6 Africa, 2/6 Europe, 3/6 Asia
    admixture_event = [
        msprime.MassMigration(
            time=Tadmix, source=3, destination=0, proportion=1.0/6.0),
        msprime.MassMigration(
            time=Tadmix+0.0001, source=3, destination=1, proportion=2.0/5.0),
        msprime.MassMigration(
            time=Tadmix+0.0002, source=3, destination=2, proportion=1.0)
    ]
```

So backwards in time, the first event is to take 1/6 of ADMIX, and merge with AFR (1/6 African). Then, 2/5 of the remainder (1-1/6 = 5/6) of ADMIX is merged into EUR (2/5 * 5/6 = 2/6 European). Then, all of the remainder (1-1/6-2/6=3/6) of ADMIX is merged into ASIA (3/6 Asian). Here is the corresponding model debug output:
```
============================
Epoch: 0 -- 12.0 generations
============================
     start     end      growth_rate |     0        1        2        3    
   -------- --------       -------- | -------- -------- -------- -------- 
0 |1.45e+04 1.45e+04              0 |     0     2.5e-05  7.8e-06     0    
1 | 3.4e+04 3.25e+04         0.0038 |  2.5e-05     0    3.11e-05     0    
2 |4.59e+04 4.33e+04         0.0048 |  7.8e-06 3.11e-05     0        0    
3 |5.47e+04   3e+04            0.05 |     0        0        0        0    

Events @ generation 12.0
   - Mass migration: Lineages moved with probability 0.16666666666666666 backwards in time with source 3 & dest 0
                     (equivalent to migration from 0 to 3 forwards in time)
==================================
Epoch: 12.0 -- 12.0001 generations
==================================
     start     end      growth_rate |     0        1        2        3    
   -------- --------       -------- | -------- -------- -------- -------- 
0 |1.45e+04 1.45e+04              0 |     0     2.5e-05  7.8e-06     0    
1 |3.25e+04 3.25e+04         0.0038 |  2.5e-05     0    3.11e-05     0    
2 |4.33e+04 4.33e+04         0.0048 |  7.8e-06 3.11e-05     0        0    
3 |  3e+04    3e+04            0.05 |     0        0        0        0    

Events @ generation 12.0001
   - Mass migration: Lineages moved with probability 0.4 backwards in time with source 3 & dest 1
                     (equivalent to migration from 1 to 3 forwards in time)
=====================================
Epoch: 12.0001 -- 12.0002 generations
=====================================
     start     end      growth_rate |     0        1        2        3    
   -------- --------       -------- | -------- -------- -------- -------- 
0 |1.45e+04 1.45e+04              0 |     0     2.5e-05  7.8e-06     0    
1 |3.25e+04 3.25e+04         0.0038 |  2.5e-05     0    3.11e-05     0    
2 |4.33e+04 4.33e+04         0.0048 |  7.8e-06 3.11e-05     0        0    
3 |  3e+04    3e+04            0.05 |     0        0        0        0

Events @ generation 12.0002
   - Mass migration: Lineages moved with probability 1.0 backwards in time with source 3 & dest 2
                     (equivalent to migration from 2 to 3 forwards in time)
===================================
Epoch: 12.0002 -- 920.0 generations
===================================
     start     end      growth_rate |     0        1        2        3    
   -------- --------       -------- | -------- -------- -------- -------- 
0 |1.45e+04 1.45e+04              0 |     0     2.5e-05  7.8e-06     0    
1 |3.25e+04 1.03e+03         0.0038 |  2.5e-05     0    3.11e-05     0    
2 |4.33e+04    554           0.0048 |  7.8e-06 3.11e-05     0        0    
3 |  3e+04  5.76e-16           0.05 |     0        0        0        0
```

The SLiM code I generate with #377 behaves the following way... A new population ADMIX is created with `N * proportion` parents from ASIA using `sim.addSubpopSplit()`. This takes effect immediately when the addSubpopSplit() call is made. Then two migration rates are set, 2/5 for EUR->ADMIX, and 1/6 for AFR->ADMIX, using `pop.setMigrationRates()`. These calls do not do anything immediate, but during the next stage of offspring generation (the start of the next generation, roughly), 2/5 of ADMIX's children are drawn from EUR, and 1/6 from AFR. In the generation after the migration rates are set, I reset migration rates to zero again.

The debug when running the SLiM code is (first column is the generation number):
```
592: {dbg(self.source); sim.addSubpopSplit(3,2999,2);}
...
592: {dbg(self.source); p3.setMigrationRates(1, 0.4);}
592: {dbg(self.source); p3.setMigrationRates(0, 0.166667);}
593: {dbg(self.source); p3.setMigrationRates(1, 0);}
593: {dbg(self.source); p3.setMigrationRates(0, 0);}
```

So, should the SLiM engine assume that `Tadmix = Tadmix + 0.0001`, *and* calculate a new migration matrix by multiplying for these distinct epochs? This seems messy. It might be slightly less messy if the time delta didn't exist. Any other suggestions?

**Comment by grahamgower** [comment]
Closed by #400.
